src/datasets/arrow_dataset.py

```diff
@@ -159,6 +159,51 @@ def version(self):
         return self._info.version
 
 
+class TensorflowDatasetMixIn:
+    def __init__(self):
+        pass
+
+    def to_tf_dataset(self, tokenizer, cols_to_remove, batch_size, shuffle):
+        import tensorflow as tf
+        dataset_in = self.remove_columns(cols_to_remove)
+        tf_cols = [col for col in dataset_in.features]
+        label_index = tf_cols.index("label")
+        dtypes_out = []
+        for col in tf_cols:
+            try:
+                col_feature = dataset_in.features[col]
+                if hasattr(col_feature, 'feature'):
+                    col_feature = col_feature.feature
+                dtype_str = col_feature.dtype
+                dtypes_out.append(tf.as_dtype(dtype_str))
+            except TypeError:
+                raise TypeError(f"Couldn't convert column {col}, dtype {dtype_str} to TF Tensor!")
+
+        def indices_to_samples(indices):
+            batch = dataset_in.select(list(indices), keep_in_memory=True).to_dict()
+            batch = tokenizer.pad(batch)
+            output = []
+            for col in tf_cols:
+                output.append(batch[col])
+            return output
+
+        def graph_indices_to_samples(indices):
+            return tf.py_function(indices_to_samples, [indices], Tout=dtypes_out)
+
+        def reform_dict(*batch_list):
+            return ({col: batch_list[i] for i, col in enumerate(tf_cols)}, batch_list[label_index])
+
+        indices = tf.range(len(dataset_in))
+        tf_dataset = tf.data.Dataset.from_tensor_slices(indices)
+        if shuffle:
+            tf_dataset = tf_dataset.shuffle(buffer_size=len(tf_dataset))
+        tf_dataset = tf_dataset.batch(batch_size)
+        tf_dataset = tf_dataset.map(graph_indices_to_samples).map(reform_dict)
```

Comment:
This looks super slick! Did you have a chance to try it on multi-GPU and TPU to check it all works okay?

Reply:
Trying that out now, but any working `tf.data.Dataset` should work on all of them!

Reply:
Performance is way better than before - we can iterate through all of fashion-mnist in about 2-3 seconds on a single core, and it can prefetch data in parallel with model execution. I can go back and try to add in multiprocessing later for people with really big training setups, but performance should be good enough for almost all use cases now.